### PR TITLE
Fixing platform_release pep440

### PR DESCRIPTION
**Pull request by npapapietro**
Resolves: python-poetry/poetry#7418
Resolves: python-poetry/poetry#9434

- [x] Added **tests** for changed code.
- [ ] Updated **documentation** for changed code.

Can add more tests if requested. Any feedback welcome as well, attempted to fix with smallest changes and no side effects.

Examples of conditions this should fix

- `'tegra' in platform_release`
- `'platform_release <= 5.10.123-tegra`
- `'arm' in platform_version`

When parsing versions, I added an additional semver-ish regex that will catch patterns that match certain `platform_release` versions that aren't quite pep440 or semver. Examples are either WSL2 builds, raspi builds or nvidia edge device builds

- `5.15.146.1-microsoft-standard-WSL2`
- `6.6.20+rpt-rpi-v8`
- `5.10.120-tegra`

### CODE REVIEW

**Comment by radoering** [comment]
Interesting approach. I'll try to take a closer look later this week.

For future reference: #552 is a previous attempt to solve the issue by just making `platform_release` "non-version-like". The comments in this PR might be interesting to understand why it is not that simple.

**Comment by npapapietro** [comment]
Because this addresses some issues for folks using edge devices (my team using `-tegra` nvidia devices), it was important to keep it version-like still. The big takeaways here are:
- Adding a string comparison flag `str_cmp` to detect when `'substring' in platform_release` for simple testing
- Since OS versions aren't semver like, we choose to cheese it a bit and extract the maj,min,patch components and then relegate the rest of build tag to the `Version` classes `post` arg.

**Review by radoering — CHANGES_REQUESTED** [review]
I think with some improvements and some more unit tests for the changed methods the PR could be accepted.

One general advice: It makes the review easier if you re-arrange your commits so that there is a corresponding unit test for each change in the same commit. For example:
- In the first commit, you changed the `validate()` method, but you did not extend a `validate` test.
- In the second commit, you changed the version constraint parser, but you did not extend `test_parse_constraint`.

If you change a method, please take a look if there is a unit test for this method and extend the test.

**Comment by npapapietro** [comment]
Thank you for your review, I will work to fixup my commit flow to match change+unit test and follow up on your comments.

File: src/poetry/core/version/markers.py [hunk 1]
```diff
@@ -449,8 +459,17 @@ def __hash__(self) -> int:
...
     def __str__(self) -> str:
+        if self._str_cmp:
+            return f'"{self._value}" {self._operator} {self._name}'
         return f'{self._name} {self._operator} "{self._value}"'
 
+    def validate(self, environment: dict[str, Any] | None) -> bool:
+        if self._operator in {"in", "not in"} and isinstance(
+            self._constraint, Constraint
+        ):
+            self._constraint = Constraint(self.value, self._operator)
```
[the remark this answers is not in the document]
Comment (npapapietro) [reply src/poetry/core/version/markers.py]: I moved it to `__init__` so no need to override `validate`

File: src/poetry/core/constraints/version/parser.py [hunk 2]
```diff
@@ -180,6 +177,39 @@ def parse_single_constraint(
...
+        minor = m.group("minor")
+        patch = m.group("patch")
+        # certain OS releases aren't semver or pep440 compliant
+        build = m.group("build")
+        try:
+            version = Version.from_parts(
+                major=int(major),
+                minor=int(minor) if minor else None,
+                patch=int(patch) if patch else None,
+                post=ReleaseTag(phase=build),
```
[the remark this answers is not in the document]
Comment (npapapietro) [reply src/poetry/core/constraints/version/parser.py]: I followed your comment below, I changed the regex to to match release+build. From then passed the release_string to Version and set `build` as the `local` component.

File: src/poetry/core/constraints/version/patterns.py [hunk 3]
```diff
@@ -5,6 +5,24 @@
...
+        \.
+        (?P<patch>0|[1-9]\d*)
+        (?P<nonstd>(?:\.(0|[1-9]\d*))*)?
+    )
+)
+(?:\+|-(?P<build>
+    [0-9a-zA-Z-]+
+    (?:\.[0-9a-zA-Z-]+)*
+))?
+"""
```
[the remark this answers is not in the document]
Comment (npapapietro) [reply src/poetry/core/constraints/version/patterns.py]: https://github.com/python-poetry/poetry-core/pull/722#discussion_r1582497206


Was able to simplify the regex to just "release" and "build".

**Comment by npapapietro** [comment]
I have forced pushed over my previous commits with 3 commits hopefully tying change+test into each commit. There might be some overlap between tests. I replied to your comments 

> If you change a method, please take a look if there is a unit test for this method and extend the test.

I added several more tests to the PR by looking at the methods I modified. Please let me know if there are ones I overlooked or additional test cases you would like to see.

File: src/poetry/core/constraints/generic/constraint.py [hunk 4]
```diff
@@ -98,13 +125,21 @@ def intersect(self, other: BaseConstraint) -> BaseConstraint:
...
+            if (
+                (other.operator == "!=" and self.operator == "==")
+                or (other.operator == "not in" and self.operator == "in")
+            ) and other.allows(self):
                 return self
 
-            if other.operator == "!=" and self.operator == "!=":
+            if (other.operator == "!=" and self.operator == "!=") or (
+                other.operator == "not in" and self.operator == "not in"
+            ):
```
Comment (radoering) [inline src/poetry/core/constraints/generic/constraint.py:142]: I think some combinations are missing here. For example, the intersection of `platform_release != "5.10.123-tegra"` and `"tegra" not in platform_release` is `"tegra" not in platform_release`.

File: src/poetry/core/constraints/generic/constraint.py [hunk 5]
```diff
@@ -118,13 +153,21 @@ def union(self, other: BaseConstraint) -> BaseConstraint:
...
+            if (
+                (other.operator == "!=" and self.operator == "==")
+                or (other.operator == "not in" and self.operator == "in")
+            ) and other.allows(self):
                 return other
 
-            if other.operator == "==" and self.operator == "==":
+            if (other.operator == "==" and self.operator == "==") or (
+                other.operator == "in" and self.operator == "in"
+            ):
```
Comment (radoering) [inline src/poetry/core/constraints/generic/constraint.py:170]: I think some combinations are missing here. For example, the union of `platform_release != "5.10.123-tegra"` and `"tegra" not in platform_release` is `platform_release != "5.10.123-tegra"`.

**Review by radoering — CHANGES_REQUESTED** [review]
Thanks for reorganizing the commits. This really makes the review easier. I added some further comments.

The hardest part is to consider all combinations of operators in the methods of `Constraint`. I mentioned some examples in the last comments, which can be added in `test_constraints` if we can handle them.

File: src/poetry/core/version/markers.py [hunk 6]
```diff
@@ -322,7 +322,10 @@ class SingleMarker(SingleMarkerLike[Union[BaseConstraint, VersionConstraint]]):
     }
 
     def __init__(
-        self, name: str, constraint: str | BaseConstraint | VersionConstraint
+        self,
+        name: str,
+        constraint: str | BaseConstraint | VersionConstraint,
+        str_cmp: bool = False,
```
[the remark this answers is not in the document]
Comment (npapapietro) [reply src/poetry/core/version/markers.py:328]: perhaps "substr_match" would work better? What does python call `substr in string` test if anything? My var name game is definitely lacking here.

File: tests/constraints/generic/test_main.py [hunk 7]
```diff
@@ -18,6 +18,8 @@
         ("==win32", Constraint("win32", "=")),
         ("!=win32", Constraint("win32", "!=")),
         ("!= win32", Constraint("win32", "!=")),
+        ("not integra", Constraint("tegra", "not in")),
+        ("integra", Constraint("tegra", "in")),
```
[the remark this answers is not in the document]
Comment (npapapietro) [reply tests/constraints/generic/test_main.py:22]: Yeah, this is how the `{op}{value}` is being generated. I was unsure about changing that method and introducing regression bugs with everything else. I'll play around with it to see if I can tweak it to increase readability without introducing many side effects

File: src/poetry/core/version/markers.py [hunk 8]
```diff
@@ -380,6 +390,15 @@ def __init__(
...
 
+        # This is a fixup on constraint if its a string like
+        # comparison for platform_release
+        if (
+            self._operator in {"in", "not in"}
+            and self._str_cmp
+            and isinstance(parsed_constraint, Constraint)
+        ):
+            parsed_constraint = Constraint(self._value, self._operator)
+
```
[the remark this answers is not in the document]
Comment (npapapietro) [reply src/poetry/core/version/markers.py:401]: I'll double check my work and see. This might be unneeded now because of other changes.

**Comment by npapapietro** [comment]
Will address comments and fix issues in coming days

File: src/poetry/core/version/markers.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (radoering) [reply src/poetry/core/version/markers.py:328]: Nah, it is not just `in`. For example, `str_cmp` is also `True` for `"2.7" == python_version`. The value of the parameter only depends on which side of the operator the name of the marker is. (And I think that makes sense, we just need another name.)

[the remark this answers is not in the document]
Comment (npapapietro) [reply src/poetry/core/version/markers.py:328]: Going with `swapped_name_value`

File: tests/constraints/generic/test_main.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (npapapietro) [reply tests/constraints/generic/test_main.py:22]: Investigated a bit further, I can without much effort space out the operator from the value `not integra` -> `not in tegra`.

Changing the order of the values would require some reworking.
This one https://github.com/npapapietro/poetry-core/blob/ddfc68da2382e4581b6c4e6b278a7be56d584d44/src/poetry/core/constraints/generic/parser.py#L31 would require increasing the complexity of the regex itself or if statements to ensure mixing `not in`/`in` operators with others are still satisfied.

File: src/poetry/core/version/markers.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (npapapietro) [reply src/poetry/core/version/markers.py:401]: This might have been a premature addition before other logic was in place, removing it

**Comment by npapapietro** [comment]
I've updated the `allows` method and tests related to `intersection` and `union`. I definitely need a sanity check. I included a logic table to double check myself.

File: tests/constraints/generic/test_main.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (radoering) [reply tests/constraints/generic/test_main.py:22]: Actually, I have no idea why this regex is so complicated. I think we can simplify it so it will not be an issue. See my other comment. I think I am still in favor of bringing it in the correct order despite some additional effort. Adapting `STR_CMP_CONSTRAINT` and `SingleMarker._CONSTRAINT_RE` should not be too complicated. In `STR_CMP_CONSTRAINT` we just have to swap `op` and `value` and in `SingleMarker._CONSTRAINT_RE` we can just add an "or" with swapped order for the new operators.

**Review by radoering — CHANGES_REQUESTED** [review]
I think the logic for `allows` is not yet completely correct. However, while reviewing it, I noticed that this method had already been quite a mess with a vague interface and some bugs before. I think we have to fix it first for the existing operators (and increase test coverage) before introducing new operators. Therefore, I created #732. I hope it does not take too long until this one gets reviewed so we can proceed here.

In the meantime, please take a look at my other comments concerning the regexes.

File: src/poetry/core/version/markers.py [hunk 9]
```diff
@@ -438,7 +448,11 @@ def invert(self) -> BaseMarker:
             # We should never go there
             raise RuntimeError(f"Invalid marker operator '{self._operator}'")
 
-        return parse_marker(f"{self._name} {operator} '{self._value}'")
+        if self._swapped_name_value:
+            constraint = f'"{self._value}" {operator} {self._name}'
+        else:
+            constraint = f"{self._name} {operator} '{self._value}'"
```
[the remark this answers is not in the document]
Comment (npapapietro) [reply src/poetry/core/version/markers.py:454]: done

**Comment by npapapietro** [comment]
I managed to make some changes to the regex and how we represent the marker internally. Now all markers should be correctly represented.

`"tegra" in platform_release` is now `"tegra" in` for the single marker instead of `integra`

**Review by radoering — CHANGES_REQUESTED** [review]
Looks good so far.

I think we can drop one regex. (See code comments.)

> `"tegra" in platform_release` is now `"tegra" in` for the single marker instead of `integra`

I wonder if we should use quotes or not because the representation of the non-swapped constraints is without quotes (e.g. `!=win32` instead of `!="win32"`). However, I understand that there is more ambiguity with `in` and `not in` so I think it is fine the way it is.

By the way, #732 has been merged into main so you can rebase (sorry for the conflicts) and we can take a closer look at the `allows` methods.

**Comment by npapapietro** [comment]
I have hopefully deconflicted correctly the PR that was merged earlier and addressed your comments on the regex.

File: src/poetry/core/constraints/generic/constraint.py [hunk 10]
```diff
@@ -43,21 +67,59 @@ def operator(self) -> str:
...
+           in   || not in |   in   |   in   | not in
+         not in || not in | not in | not in | false
+
+        """
+
+        if not isinstance(other, Constraint) or other.operator not in {
+            "==",
+            "in",
+            "not in",
+        }:
```
Comment (radoering) [inline src/poetry/core/constraints/generic/constraint.py:85]: For `allows()`, `other` must be an exact value, in other words, the operator must be `==`. This will not change with `in` and `not in`.

I assume you can move part of your logic to `allows_all()` and `allows_any()` were `other` can have any operator. Further, you have to replace some `allows()` in `difference()`, `intersect()` and `union()` with `allows_all()` or `allows_any()`.

**Review by radoering — CHANGES_REQUESTED** [review]

**Comment by npapapietro** [comment]
I was traveling for work, but now I am getting back to this. I will be asking some clarification questions as I work through you latest comments.

File: src/poetry/core/constraints/generic/constraint.py (hunk 10 above)
Comment (npapapietro) [reply src/poetry/core/constraints/generic/constraint.py:85]: I can revert `allows()` to only consider `other.operator == "=="`, but I have some questions related how you would like me to implement `allows_any()` and `allows_all()` inside `difference()`, `instersect()` and `union()`.

I have searched the code base for uses of `allows_any()` and `allows_all()`, but I have found no uses outside of unit testing. I would like some clarification on usage, as well as confirmation of logic on `intersect()` and `union()` for some operators.

First some examples
|                        |                        |    Intersect      |       Union        |
|----------------|---------------|-----------------|----------------|
| `"tegra" not in`| `"tegra" not in`| `"tegra" not in` | `"tegra" not in` |
| `"tegra" in`      | `"tegra" not in`| empty set         |  all/any set       |
| `"tegra" in`      | `"tegra" in`      | `"tegra" in`        | `"tegra" in`       |
| `"teg" in`         | `"tegra" in`      | `"tegra" in`        | `"teg" in`          |
| `"teg" not in`   | `"tegra" in`      | empty set         | empty set                                               |
| `"tegra" in`      | `"rpi" in`            | empty set       | MultiConstraint( `"tegra" in`,`"rpi" in` )      |
| `!=1.2.3-tegra`   | `"tegra" in`    | `!=1.2.3-tegra`    | MultiConstraint( `!=1.2.3-tegra` ,`"tegra" in` )      |
| `!=1.2.3-tegra`   | `"tegra" not in`    | empty set         |  empty set        |

Comment (radoering) [reply src/poetry/core/constraints/generic/constraint.py:85]: I edited your table directly because I think/hope it's easier to keep track. First of all, this is a finicky topic so I might have made mistakes. If you think that some cell is wrong, please tell me. I crossed out mistakes instead of just replacing them, so you can follow the changes. If you agree feel free to remove the crossed out text.

Some explanations:

- I added a line with `"tegra" in` and `"teg" in` because order is relevant for `allows_all` in this case. (`"teg" in` allows all `"tegra" in`, but `"tegra" in` does not allow all `"teg" in`)
- I think you mixed up `MultiConstraint` and `UnionConstraint` a bit. (`MultiConstraint` is like `and` (intersection) and `UnionConstraint` is like `or` (union).)

Some general rules (I hope they are correct 😉):
- If `c1.allows_all(c2)` is `True`, `c1.allows_any(c2)` must also be `True`.
- If `c1.allows_any(c2)` is `False`, `c1.allows_all(c2)` must also be `False`.
- If `c1.allows_any(c2)` is `True`, `c2.allows_any(c1)` must also be `True`.
- If the intersection is empty, `allows_any` is `False` (for both orders). If `allows_any` is `False` for both orders, the intersection is empty.
- If the intersection is not empty, `allows_any` is `True` (for both orders). If `allows_any` is `True` for both orders, the intersection is not empty.

**Comment by npapapietro** [comment]
I want to get some more test cases in to ensure I have covered edge cases. This topic is indeed finicky

File: src/poetry/core/constraints/generic/constraint.py (hunk 10 above)
Comment (npapapietro) [reply src/poetry/core/constraints/generic/constraint.py]: After reviewing and the code, I think there might be some ambiguity around 

|  constraint 1  |  constraint 2 |    Intersect      |       Union        |   allow_any |       allow_all  |
|----------------|---------------|-----------------|----------------|-----------------|----------------| 
| `!=1.2.3-tegra`   | `"tegra" in`    |   MultiConstraint( `!=1.2.3-tegra` ,`"tegra" in` )   |   all/any set    | `True` | `False` |

I think the union should be `UnionConstraint` instead of `AnyConstraint`

Comment (radoering) [reply src/poetry/core/constraints/generic/constraint.py]: Can you give an example that is not allowed by `!=1.2.3-tegra or "tegra" in`? (If the union allows all possible values, then it is `AnyConstraint`.)

Comment (npapapietro) [reply src/poetry/core/constraints/generic/constraint.py]: I cannot, I think I was able to fix all the cases above

**Comment by radoering** [comment]
I added a few more tests, fixed and simplified a few things. I think I will take another look next week and - if everything still makes sense to me - finally merge it.

Thanks for your endurance.

**Review by radoering — APPROVED** [review]
